compiler/rustc_borrowck/src/diagnostics/mod.rs

```diff
@@ -1131,13 +1134,19 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                                 place_name, partially_str, loop_message
                             ),
                         );
-                        let ty = tcx.erase_regions(moved_place.ty(self.body, self.infcx.tcx).ty);
+                        let infcx = tcx.infer_ctxt().build();
+                        let ty = infcx.freshen(moved_place.ty(self.body, tcx).ty);
                         if let ty::Adt(def, substs) = ty.kind()
-                            && Some(def.did()) == self.infcx.tcx.lang_items().pin_type()
+                            && Some(def.did()) == tcx.lang_items().pin_type()
                             && let ty::Ref(_, _, hir::Mutability::Mut) = substs.type_at(0).kind()
-                            // FIXME: this is a hack because we can't call `can_eq`
-                            && ty.to_string() ==
-                                tcx.fn_sig(method_did).input(0).skip_binder().to_string()
+                            && let self_ty = infcx.freshen(
```

Comment:
Is this freshen call needed? It shouldn't be, since this type shouldn't reference any regions from the old inference context... :thinking: